src/overlays/actors/ovl_En_Fr/z_en_fr.c

```diff
@@ -27,99 +183,1196 @@ const ActorInit En_Fr_InitVars = {
+        this->frogSongTimer = 100;
+        func_80106CCC(globalCtx);
+        this->actionFunc = EnFr_SetReward;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1B638.s")
+void EnFr_SetReward(EnFr* this, GlobalContext* globalCtx) {
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1B778.s")
+    u16 songIndex;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1B8D0.s")
+    sEnFrPointers.flags = 0xC;
+    songIndex = this->songIndex;
+    this->actionFunc = EnFr_Deactivate;
+    this->reward = 0;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1B9D8.s")
+    // 5 Child Song Rewards: 1st time Purple Rupee, afterwards blupee
+    if ((songIndex >= 0) && (songIndex < 5)) {
+        if (!(gSaveContext.eventChkInf[13] & sSongIndex[songIndex])) {
+            gSaveContext.eventChkInf[13] |= sSongIndex[songIndex];
+            this->reward = GI_RUPEE_PURPLE;
+        } else {
+            this->reward = GI_RUPEE_BLUE;
+        }
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1BB34.s")
+    // Song of Storms Reward: 1st time HP, afterwards blupee
+    else if (songIndex == 5) {
+        if (!(gSaveContext.eventChkInf[13] & sSongIndex[songIndex])) {
+            gSaveContext.eventChkInf[13] |= sSongIndex[songIndex];
+            this->reward = GI_HEART_PIECE;
+        } else {
+            this->reward = GI_RUPEE_BLUE;
+        }
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1BC14.s")
+    // Frogs 2 Reward: 1st time HP, afterwards purple rupee
+    else if (songIndex == 6) {
+        if (!(gSaveContext.eventChkInf[13] & sSongIndex[songIndex])) {
+            gSaveContext.eventChkInf[13] |= sSongIndex[songIndex];
+            this->reward = GI_HEART_PIECE;
+        } else {
+            this->reward = GI_RUPEE_PURPLE;
+        }
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1BCB4.s")
+void EnFr_Deactivate(EnFr* this, GlobalContext* globalCtx) {
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1BD94.s")
+    EnFr* frog_loop1;
+    EnFr* frog_loop2;
+    int frogIndex;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1BDFC.s")
+    // Originally was going to have separate butterfly actor
+    // Changed to include butterfly as part of frog actor
+    // This unused code would have frozen the butterfly actor above frog
+    if (this->unusedButterflyActor) {
+        this->unusedButterflyActor->freezeTimer = 0xA;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1BE98.s")
+    for (frogIndex = 0; frogIndex < 5; frogIndex++) {
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1BF10.s")
+        frog_loop1 = sEnFrPointers.frogs[frogIndex];
+        // Translation: There are no frogs!?
+        if (!frog_loop1) {
```

Comment:
```suggestion
        if (frog_loop1 == NULL) {
```
Pointers should be compared to NULL. There's a few more of these in the file.